File: feiqiu/src/lib/types/conversations.ts
/**
 * Conversation Types
 *
 * Types for conversations and conversation participants.
 */

import type { ConversationType } from "./messaging";

/**
 * Conversation Participant (from backend)
 */
export interface ConversationParticipant {
  id: number;
  conversationId: number;
  peerIp: string;
  joinedAt: number;
  leftAt: number | null;
  role: string;
}

/**
 * Conversation DTO (from backend)
 */
export interface ConversationDto {
  id: number;
  type: ConversationType;
  createdAt: number;
  updatedAt: number;
  isPinned: boolean;
  isArchived: boolean;
  isMuted: boolean;
  unreadCount: number;
  lastMessageId: number | null;
  lastMessageAt: number | null;
  lastMessageContent: string | null;
  lastMessageType: string | null;
  participants: ConversationParticipant[];
}

/**
 * Input for updating a conversation
 */
export interface UpdateConversationInput {
  id: number;
  isPinned?: boolean;
  isArchived?: boolean;
  isMuted?: boolean;
}

File: feiqiu/src-tauri/src/storage/entities/conversations.rs
// src-tauri/src/storage/entities/conversations.rs
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "conversations")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = true)]
    pub id: i32,

    /// Conversation type: 'single' or 'group'
    #[sea_orm(column_type = "Text")]
    pub r#type: String,

    #[sea_orm(column_type = "BigInteger")]
    pub created_at: DateTime,

    #[sea_orm(column_type = "BigInteger")]
    pub updated_at: DateTime,

    #[sea_orm(column_type = "Boolean", default_value = "false")]
    pub is_pinned: bool,

    #[sea_orm(column_type = "Boolean", default_value = "false")]
    pub is_archived: bool,

    #[sea_orm(column_type = "Boolean", default_value = "false")]
    pub is_muted: bool,

    #[sea_orm(column_type = "Integer", default_value = "0")]
    pub unread_count: i32,

    #[sea_orm(column_type = "Integer", nullable)]
    pub last_message_id: Option<i32>,

    #[sea_orm(column_type = "BigInteger", nullable)]
    pub last_message_at: Option<DateTime>,

    #[sea_orm(column_type = "Text", nullable)]
    pub last_message_content: Option<String>,

    #[sea_orm(column_type = "Text", nullable)]
    pub last_message_type: Option<String>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        has_many = "super::conversation_participants::Entity"
    )]
    ConversationParticipants,
}

impl Related<super::conversation_participants::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::ConversationParticipants.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
